### ISSUE TITLE
Assertion Error when running NAS experiment

### ISSUE THREAD

**Issue by lrlofgren** [microsoft/nni#4732]
I am trying to run an NAS experiment and I get an assertion error, and cannot find anything about this issue. The error is pasted below:

File D:\Data\AlgorithmDevelopment\April04_2022\NAS_20220404.py:326 in <module>
    exp = RetiariiExperiment(BaseModel, model_evaluator, [], search_strategy)

  File ~\.conda\envs\spyder\lib\site-packages\nni\retiarii\experiment\pytorch.py:183 in __init__
    self._dispatcher = RetiariiAdvisor()

  File ~\.conda\envs\spyder\lib\site-packages\nni\retiarii\integration.py:52 in __init__
    register_advisor(self)  # register the current advisor as the "global only" advisor

  File ~\.conda\envs\spyder\lib\site-packages\nni\retiarii\integration_api.py:23 in register_advisor
    assert _advisor is None

AssertionError

**Environment**:
- NNI version:
- Training service (local|remote|pai|aml|etc): local
- Client OS: Windows 10
- Python version:
- PyTorch/TensorFlow version:
- Is conda/virtualenv/venv used?: conda

**Comment by lrlofgren** [comment]
More context: I start the experiment and it will run a trial, then will give me the error posted above after the first trial is complete.

**Comment by matluster** [comment]
Did you instantiate two `RetiariiExperiment` in one process?

**Comment by lrlofgren** [comment]
I do not believe so.  I have posted my code below. However, I would like to add that this error also occurs when I run the MNIST search.py program. 


@model_wrapper
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        
        self.hidden_dim_1 = nn.ValueChoice([5, 10, 15, 20, 25, 30, 40, 50, 60, 70, 80, 90, 100, 256, 512])

        self.fc1 = nn.Linear(34,self.hidden_dim_1,bias = True)
           
        self.InRepeat = nn.Sequential(
            nn.Linear(in_features=self.hidden_dim_1, out_features=self.hidden_dim_1, bias=True),
            nn.LeakyReLU(inplace=True))
        
        self.Repeat = nn.Sequential(nn.Repeat(self.InRepeat,[1,10],label='Repeated Hidden Layers'))
        
        self.fc3 = nn.Linear(self.hidden_dim_1, 1)
        
    def forward(self, x):
        
        x = F.leaky_relu(self.fc1(x))
        x = self.Repeat(x)
        output = self.fc3(x)
        
        return output

############################################################################### 
############################################################################### 

def train(model, train_loader, device, optimizer, epoch):
    
    model.train()
    running_loss = 0.
    total_n = 0
    
    for batch_idx, (data, target) in enumerate(train_loader):
        data = data.to(device)
        target = target.to(device)

        #zero gradients for every batch
        optimizer.zero_grad()
        #make predictions for this batch
        output = model(data)
        #compute loss and gradients
        loss = F.binary_cross_entropy_with_logits(output, target)
        running_loss += loss.item() * len(data)
        total_n += len(data)
        loss.backward()
        
        #adjust weights
        optimizer.step()

    epoch_loss = running_loss/total_n
    return epoch_loss      

###############################################################################    
############################################################################### 

def test_f1_auc(model, device, test_loader):
    
    model.eval()
    sig = nn.Sigmoid()
    test_loss = 0
    with torch.no_grad():
        
        for data, target in test_loader:

            data = data.to(device)
            target = target.to(device)
            # make predictions
            output = model(data)
            # sum up batch loss
            test_loss += F.binary_cross_entropy_with_logits(output, target).item()
            # get the index of the max log-probability
            pred_prob = sig(output)
            pred = np.round(sig(output).cpu().detach())
            
            fpr, tpr, thresholds = roc_curve(target.cpu().detach().numpy(), pred_prob.view_as(target).cpu().detach().numpy(), pos_label=1)
            opt_ind = np.argmax(tpr - fpr)
            opt_thresh = thresholds[opt_ind]
            auc_prob = auc(fpr, tpr)
            predicted = (pred_prob > opt_thresh) * 1

            score = f1_score(target.cpu().detach().numpy() , pred.view_as(target).cpu().detach().numpy(), pos_label = 1 )
            adj_score = f1_score(target.cpu().detach().numpy() , predicted.view_as(target).cpu().detach().numpy(), pos_label = 1 )

    return score, adj_score, auc_prob

############################################################################### 
############################################################################### 

def reset_weights(m):
    for layer in m.children():
        if hasattr(layer, 'reset_parameters'):
            layer.reset_parameters()

############################################################################### 
###############################################################################
            
class InpLabDataset(Dataset):
    
    def __init__(self, inputs, labels):

        self.inputs = torch.from_numpy(inputs)
        self.labels = torch.from_numpy(labels)

        self.groups = np.unique(labels)
        self.n_samples = self.inputs.shape[0]

    def __getitem__(self,index):
        return self.inputs[index], self.labels[index]
        
    def __len__(self):
        return self.n_samples
    
    def __groups__(self):
        return self.groups

############################################################################### 
############################################################################### 

def evaluate_model(Net):
    
    model = Net()

    params = {
        'batch_size': 8,
        'learning_rate': 0.00001,
        'momentum':0.5,
        'epochs':100,
        }
    
    optimizer = torch.optim.Adam(model.parameters(), lr=params['learning_rate'])
    start_time1 = t.time()
    torch.manual_seed(42)


    # prepare dataset, splitting into x and y

    # n_classes = len(total_dataset.groupby(['TOTAL_AKI']))
    # n_features = x_dataset.shape[1]

    # creating dataset
    dataset = InpLabDataset(x_dataset, y_dataset)
    
    # splitting into 3 folds
    k_folds = 3
    kfold = StratifiedKFold(n_splits=k_folds, shuffle=True, random_state = 42)

    # build network

    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
    model.to(device)
    results = {}
    best_results = {}
    best_epoch_n = {}
    
    auc_results = {}
    best_auc_results = {}
    auc_epoch_n = {}
    
    adj_results = {}
    best_adj_results = {}
    best_adj_epoch_n = {}
    
    print(" --- %s seconds ---" %(t.time() - start_time1))

    for fold, (train_idx, test_idx) in enumerate(kfold.split(x_dataset, y_dataset)):
        torch.manual_seed(42)
        train_subsampler = torch.utils.data.SubsetRandomSampler(train_idx)
        test_subsampler = torch.utils.data.SubsetRandomSampler(test_idx)
        trainloader = torch.utils.data.DataLoader(dataset, batch_size=params['batch_size'], sampler=train_subsampler, pin_memory=True)
    
        trainloader_test = torch.utils.data.DataLoader(dataset, batch_size=len(train_subsampler), sampler=train_subsampler, pin_memory=True)
        testloader = torch.utils.data.DataLoader(dataset, batch_size=len(test_subsampler), sampler=test_subsampler, pin_memory=True)
        
        model.apply(reset_weights)
       
        best = 0.
        best_auc = 0.
        best_adj = 0.
        
        loss_list = []
        testing_score = []
        epoch_10_list = []

        for epoch in range(1, params['epochs'] + 1):

            l = train(model, trainloader, device, optimizer, epoch)
            loss_list.append(l)
            # epoch_loss.append(l)
            if epoch % 10 == 0:
                int_test_score, adj_int_score, int_auc = test_f1_auc(model, device, testloader)
                testing_score.append(int_test_score)
                epoch_10_list.append(epoch)
                
                if int_test_score > best:
                    best_results[str(fold)] = int_test_score
                    best_epoch_n[str(fold)] = epoch
                    best = int_test_score
                    
                if int_auc > best_auc:
                    best_auc_results[str(fold)] = int_auc
                    auc_epoch_n[str(fold)] = epoch
                    best_auc = int_auc
                    
                if adj_int_score > best_adj:
                    best_adj_results[str(fold)] = adj_int_score
                    best_adj_epoch_n[str(fold)] = epoch
                    best_adj = adj_int_score
                    
        test_score, adj_test_score, auc = test_f1_auc(model, device, testloader)
        train_score, adj_train_score, train_auc = test_f1_auc(model, device, trainloader_test)
        
        auc_results[str(fold)] = auc
        results[str(fold)] = test_score
        adj_results[str(fold)] = adj_test_score
        nni.report_intermediate_result(auc)
    
    total = 0.
    for val in results.values():
        total += val
        
    avg = total / len(results)
    
    nni.report_final_result(avg)


if __name__ == '__main__':
    base_model = Net()

    search_strategy = strategy.TPEStrategy()
    model_evaluator = FunctionalEvaluator(evaluate_model)
 
    exp = RetiariiExperiment(base_model, model_evaluator, [], search_strategy)
    
    exp_config = RetiariiExeConfig('local')
    exp_config.experiment_name = 'NAS_Search'
    exp_config.trial_concurrency = 2
    exp_config.max_trial_number = 10
    exp_config.training_service.use_active_gpu = True
    exp_config.trial_gpu_number = 1
    export_formatter = 'dict'
    
    exp.run(exp_config, 8081 + random.randint(0, 100))

    print('Final Model:')
    for model_code in exp.export_top_models(formatter=export_formatter):
        print(model_code)

**Comment by lrlofgren** [comment]
I will also add that the problem seems to be solved when I include exp_config.execution_engine = 'base'; however from my understanding, this is not encouraged and I should be able to use the pure python execution engine.

**Comment by lrlofgren** [comment]
Just read somewhere else that i should be using WSL. is this true?

**Comment by matluster** [comment]
Sorry but your code looks a little too messy. From a first glance, I didn't see any obvious problem.

You mentioned that you also encountered this problem when running MNIST example. Could you post your console log, as well as nnimanager.log (you can find it in `~/nni-experiments`)?

You can surely try WSL, but I'm not confident that this issue is related to Windows.
